File: packages/java-pom-bdd/src/test/java/com/example/pages/RegisterPage.java
package com.example.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

public class RegisterPage extends BasePage {
    private final Locator dialog;
    private final Locator usernameInput;
    private final Locator emailInput;
    private final Locator passwordInput;
    private final Locator confirmPasswordInput;
    private final Locator registerButton;

    public RegisterPage(Page page) {
        super(page);
        this.dialog = page.getByRole(com.microsoft.playwright.options.AriaRole.DIALOG,
                new Page.GetByRoleOptions().setName("新規登録"));
        this.usernameInput = dialog.getByLabel("ユーザー名:");
        this.emailInput = dialog.getByLabel("メールアドレス");
        this.passwordInput = dialog.getByLabel("パスワード:");
        this.confirmPasswordInput = dialog.getByLabel("パスワード確認:");
        this.registerButton = dialog.getByRole(com.microsoft.playwright.options.AriaRole.BUTTON,
                new Locator.GetByRoleOptions().setName("新規登録"));
    }

    public boolean isVisible() {
        return dialog.isVisible();
    }

    public void register(String username, String email, String password) {
        usernameInput.fill(username);
        emailInput.fill(email);
        passwordInput.fill(password);
        confirmPasswordInput.fill(password);
        registerButton.click();
    }

    public void registerWithDifferentPasswords(String username, String email, String password, String confirmPassword) {
        usernameInput.fill(username);
        emailInput.fill(email);
        passwordInput.fill(password);
        confirmPasswordInput.fill(confirmPassword);
        registerButton.click();
    }
}

File: packages/ts-pom-bdd-di/pages/RegisterPage.ts
import { Page, Locator } from '@playwright/test';

export class RegisterPage {
  private readonly dialog: Locator;
  private readonly usernameInput: Locator;
  private readonly emailInput: Locator;
  private readonly passwordInput: Locator;
  private readonly confirmPasswordInput: Locator;
  private readonly registerButton: Locator;
  private readonly closeButton: Locator;
  private readonly loginLink: Locator;
  private readonly errorMessage: Locator;

  constructor(private page: Page) {
    this.dialog = page.getByRole('dialog', { name: '新規登録' });
    this.usernameInput = this.dialog.getByLabel('ユーザー名:');
    this.emailInput = this.dialog.getByLabel('メールアドレス');
    this.passwordInput = this.dialog.getByLabel('パスワード:').first();
    this.confirmPasswordInput = this.dialog.getByLabel('パスワード確認:');
    this.registerButton = this.dialog.getByRole('button', { name: '新規登録' });
    this.closeButton = this.dialog.getByRole('button', { name: '閉じる' });
    this.loginLink = this.dialog.getByRole('link', { name: 'ログイン' });
    this.errorMessage = this.dialog.locator('[class*="error"]');
  }

  async isVisible() {
    return await this.dialog.isVisible();
  }

  async register(username: string, email: string, password: string, confirmPassword: string) {
    await this.usernameInput.fill(username);
    await this.emailInput.fill(email);
    await this.passwordInput.fill(password);
    await this.confirmPasswordInput.fill(confirmPassword);
    await this.registerButton.click();
  }

  async close() {
    await this.closeButton.click();
  }

  async clickLoginLink() {
    await this.loginLink.click();
  }

  async getErrorMessage() {
    if (await this.errorMessage.isVisible()) {
      return await this.errorMessage.textContent();
    }
    return null;
  }

  async hasError() {
    return await this.errorMessage.isVisible();
  }
}
